### PR TITLE
Dynamic load generation with remote control

### PR DESCRIPTION
**Pull request by cachedout**
## What
This PR introduces a HTTP-based control plane for the Opbeans Load Generator

## Why
This is a part of a series of enhancements being made to the APM Integration Test Suite which collectively introduce the ability to dynamically constrain various components (such as CPU/memory/network) for the Opbeans and also to dynamically apply and modify load which is applied to those Opbeans.

This PR contains the bulks of the needed changes the load-generation service to allow load-generation characteristics to be controlled via HTTP requests to a listening server.

## How
Much of the approach is described in the in-line documentation, but at a high-level, the change being made here is to add a web server which, in turn, can control the characteristics of the Molotov process. Please see the various READMEs for additional details.

## Caveats
This new set of features is designed for use with upcoming changes to the APM Integration Test suite. As such, there is some some documentation which references yet-to-be-merged features.

### CODE REVIEW

**Review by mdelapenya — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -2,28 +2,78 @@
...
+
+### Running in HTTP mode in a container [Recommended]
+To run the Opbeans Load Generator in HTTP mode, start the container by defining the environment variable `WS` in the context of the running container. Typically, this is done using the `-e` flag [as shown in the Docker documentation](https://docs.docker.com/compose/environment-variables/#pass-environment-variables-to-containers).
+
+#### Note on running in the APM Integration Test Suite
+When running the Opbeans Load Generator in the APM Integration Test suite with the `--dyno` flag, the `WS` environment variable will be defined automatically and the container will start in HTTP mode.
+
+### HTTP method overview
+
+The following is a list of HTTP methods and their purpose. For specifics including possible
```
Comment (v1v) [inline README.md:33]: I'd say to create a link to the docs in the dyno/README.md, something like:

```suggestion
The list of HTTP methods and their purpose can be found in the [http method overview section](dyno/README.md#http-method-overview).
```

File: dyno/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+Custom scenarios can include dynamic variables via the inclusion of certain environment variables which can control their behavior.
+
+### Controlling scenario behavior
+The following environment variables are supported inside scenarios:
+
+Environment Variable|Description
+--------------------|-----------
+`SERVER_URL`|The target for load traffic
+`SERVICE_NAME`|The type of service. Used to differenate between the specific endpoints available in different opebeans. Ex: `opbeans-python`|`default`
```
Comment (v1v) [inline dyno/README.md:46]:
```suggestion
`SERVICE_NAME`|The type of service. Used to differentiate between the specific endpoints available in different opbeans. Ex: `opbeans-python`|`default`
```

File: tests/functional/test_procfile.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                ],
+            stdout=subprocess.PIPE
+            )
+
+    assert proc.stdout == b'python: OPBEANS_BASE_URL=http://opbeans-python:3000 OPBEANS_NAME=opbeans-python molotov -v --duration 500 --delay 0.120 --uvloop molotov_scenarios.py\n'
+
+def test_entrypoint_execution(default_environment):
+    """
+    GIVEN a call to the entrypoint script
+    WHEN the enviornment variables are set for URLS, RPMS, and RLS
```
Comment (v1v) [inline tests/functional/test_procfile.py:53]:
```suggestion
    WHEN the environment variables are set for URLS, RPMS, and RLS
```

File: tests/unit/test_dyno.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        {
+            'port': '990',
+            'duration': '991',
+            'delay': '992',
+            'workers': '993',
+            'scenario': 'fake_scenario',
+            'error_weight':' 994'
+            }
+        )
+    assert socketio_mock.caled_with('service_state', {'data': {'python': 'start'}})
```
Comment (v1v) [inline tests/unit/test_dyno.py:161]:
```suggestion
    assert socketio_mock.called_with('service_state', {'data': {'python': 'start'}})
```

**Review by v1v — COMMENTED** [review]
💯 I'm still reviewing it, I'll be checkout the code locally to play a bit with it, besides I'll create a docker image and run the APM-ITs locally to validate the load-gen works as expected. 

OTOH, since this component is consumed by the APM-ITs, so we might need to add some special stage to the CI pipeline to validate changes will work as expected in the APM-ITs on a PR basis for the shake of mind, what do you think?

**Comment by cachedout** [comment]
> OTOH, since this component is consumed by the APM-ITs, so we might need to add some special stage to the CI pipeline to validate changes will work as expected in the APM-ITs on a PR basis for the shake of mind, what do you think?

Absolutely! The way I'm thinking about this is to write a series of full integration tests that live in the APM integration tests suite and validate that the API for the load-generator operates as it expects. I haven't yet begun those, but once they're complete, I think it makes a lot of sense to create a coupling between tests which run on any changes to this project and the APM Integration Tests project. 💯

File: tests/unit/test_dyno.py (hunk 4 above)
Comment (cachedout) [reply tests/unit/test_dyno.py:161]: Yikes. Thanks!

**Comment by v1v** [comment]
For clarity, I took the liberty of, initially, including @beniwohli as a reviewer, since he worked on this project, but somehow GitHub didn't show him, so I decided to include the `apm-agent-python` team.

File: Makefile [hunk 5]
```diff
@@ -15,7 +17,13 @@ build: ## Build docker image
 bats: ## Install bats in the project itself
 	@git clone https://github.com/sstephenson/bats.git
 
-prepare-test: bats ## Prepare the bats dependencies
+venv: requirements-dev.txt
```
Comment (v1v) [inline Makefile:20]: I'm running this locally, but somehow when running this goal, the dependencies are failing with some issues related to this platform is not supported, in my case I'm using MacOS

```
  xcrun -sdk macosx clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/Headers -arch arm64 -arch x86_64 -I/Users/vmartinez/work/src/github.com/elastic/opbeans-loadgen/venv/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/include/python3.8 -I/private/var/folders/dg/wfl7w2g90vz9j9fd_56858jh0000gn/T/pip-install-5dhdv67e/uvloop/vendor/libuv/include -c uvloop/loop.c -o build/temp.macosx-10.14.6-x86_64-3.8/uvloop/loop.o -O2
  In file included from uvloop/loop.c:20:
  In file included from /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/include/python3.8/Python.h:11:
  In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/limits.h:21:
  In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/limits.h:63:
  /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
```

See the log output -> https://gist.github.com/v1v/1486c9ebd87f2fb6ff6596de182d8572

Comment (cachedout) [reply Makefile:20]: Looks like you are on 3.8.  Try Python 3.6 if you're on OS X.

**Comment by beniwohli** [comment]
I'll have a closer review soon, but did you consider to use a load generator that already provides a frontend (e.g. https://locust.io/) instead of building our own for Molotov? (might be a bit late for this input, unfortunately I didn't see this PR before).

**Comment by cachedout** [comment]
> I'll have a closer review soon, but did you consider to use a load generator that already provides a frontend (e.g. https://locust.io/) instead of building our own for Molotov? (might be a bit late for this input, unfortunately I didn't see this PR before).

I like Locust a lot and actually used it recently with the load-generation for the APM Java Agent project. 

I did consider using Locust and even built a partial PoC with it but discarded it in favor of this approach. Specifically, I liked that I could use Molotov to stream pseudo-realtime stats to Redis which I'm making heavy use of. Additionally, we already have the scenarios written which was an advantage. Finally, I wanted a layer that could include socketIO streaming (which I'm consuming but not yet using) and I would have had to build that on top of Locust which seemed a bit cumbersome.

I'm open to reviewing these decisions of course! Thanks, @beniwohli

**Comment by v1v** [comment]
For the APM-ITs validation I ran the below commands:

* build opbeans-loadgen locally
```
➜  opbeans-loadgen git:(dyno) make test                                 
➜  opbeans-loadgen git:(dyno) docker images | grep opbeans/opbeans-loadgen
opbeans/opbeans-loadgen                                latest                      8b40f782c4b0        44 minutes ago       163MB
```

* Prepare the APM-its services:
```
➜  apm-integration-testing git:(master) ✗ python3 scripts/compose.py start 8.0.0 \
--apm-server-build https://github.com/elastic/apm-server.git@master \
--build-parallel \
--go-agent-version='master' \
--no-apm-server-dashboards \
 --no-apm-server-self-instrument \
 --no-kibana \
--with-agent-go-net-http \
--with-opbeans-go \
--force-build \
--no-xpack-secure \
--apm-log-level=debug
```

* Show what docker containers have been started:
```
➜  apm-integration-testing git:(master) ✗ docker ps                                   
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED             STATUS                   PORTS                                                                                                        NAMES
3967af13150e        opbeans/opbeans-loadgen:latest                                 "/app/entrypoint.sh …"   2 minutes ago       Up 2 minutes                                                                                                                          localtesting_8.0.0_opbeans-load-generator
1078069e1034        apm-integration-testing_opbeans-go                             "/opbeans-go -log-js…"   2 minutes ago       Up 2 minutes (healthy)   127.0.0.1:3003->3000/tcp                                                                                     localtesting_8.0.0_opbeans-go
d7fdf03db6a0        apm-integration-testing_agent-go-net-http                      "/testapp"               2 minutes ago       Up 2 minutes (healthy)   127.0.0.1:8080->8080/tcp                                                                                     gonethttpapp
9de5aa9efc50        apm-integration-testing_apm-server                             "/usr/bin/tini -- /u…"   2 minutes ago       Up 2 minutes (healthy)   127.0.0.1:6060->6060/tcp, 127.0.0.1:8200->8200/tcp, 127.0.0.1:14250->14250/tcp, 127.0.0.1:14268->14268/tcp   localtesting_8.0.0_apm-server
01a9fa14bbdc        postgres:10                                                    "docker-entrypoint.s…"   3 minutes ago       Up 3 minutes (healthy)   0.0.0.0:5432->5432/tcp                                                                                       localtesting_8.0.0_postgres
16f381fc892a        redis:4                                                        "docker-entrypoint.s…"   3 minutes ago       Up 3 minutes (healthy)   0.0.0.0:6379->6379/tcp                                                                                       localtesting_8.0.0_redis
e3c598239659        docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT   "/bin/tini -- /usr/l…"   26 minutes ago      Up 3 minutes (healthy)   127.0.0.1:9200->9200/tcp, 9300/tcp                                                                           localtesting_8.0.0_elasticsearch
```

* And then run the tests:

```
➜  apm-integration-testing git:(master) ✗ TARGET=test-agent-go make dockerized-test
...
...

pytest tests/agent/test_go.py --reruns 3 --reruns-delay 5 -v -s --junitxml tests/results/agent-go-junit.xml
============================= test session starts ==============================
platform linux -- Python 3.7.9, pytest-5.4.3, py-1.5.2, pluggy-0.13.1 -- /app/venv/bin/python3
cachedir: .pytest_cache
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
sensitiveurl: .*
metadata: {'Python': '3.7.9', 'Platform': 'Linux-4.19.76-linuxkit-x86_64-with-debian-10.7', 'Packages': {'pytest': '5.4.3', 'py': '1.5.2', 'pluggy': '0.13.1'}, 'Plugins': {'random-order': '1.0.4', 'html': '2.1.1', 'variables': '1.9.0', 'selenium': '1.17.0', 'base-url': '1.4.2', 'rerunfailures': '9.0', 'metadata': '1.10.0'}, 'Base URL': '', 'Driver': None, 'Capabilities': {}}
rootdir: /app, inifile: pytest.ini
plugins: random-order-1.0.4, html-2.1.1, variables-1.9.0, selenium-1.17.0, base-url-1.4.2, rerunfailures-9.0, metadata-1.10.0
collecting ... collected 4 items

tests/agent/test_go.py::test_req_go_nethttp RERUN
tests/agent/test_go.py::test_req_go_nethttp RERUN
tests/agent/test_go.py::test_req_go_nethttp RERUN
tests/agent/test_go.py::test_req_go_nethttp PASSED
tests/agent/test_go.py::test_go_nethttp_error PASSED
tests/agent/test_go.py::test_concurrent_req_go_nethttp {'node': {'name': 'd7fdf03db6a0f38464b1652f48989e1e413c41b897473fec9404cb97cd1cf337'}, 'name': 'gonethttpapp', 'runtime': {'name': 'gc', 'version': 'go1.15.6'}, 'language': {'name': 'go', 'version': 'go1.15.6'}}
{'node': {'name': 'd7fdf03db6a0f38464b1652f48989e1e413c41b897473fec9404cb97cd1cf337'}, 'name': 'gonethttpapp', 'runtime': {'name': 'gc', 'version': 'go1.15.6'}, 'language': {'name': 'go', 'version': 'go1.15.6'}}
PASSED
tests/agent/test_go.py::test_concurrent_req_go_nethttp_foobar {'node': {'name': 'd7fdf03db6a0f38464b1652f48989e1e413c41b897473fec9404cb97cd1cf337'}, 'name': 'gonethttpapp', 'runtime': {'name': 'gc', 'version': 'go1.15.6'}, 'language': {'name': 'go', 'version': 'go1.15.6'}}
{'node': {'name': 'd7fdf03db6a0f38464b1652f48989e1e413c41b897473fec9404cb97cd1cf337'}, 'name': 'gonethttpapp', 'runtime': {'name': 'gc', 'version': 'go1.15.6'}, 'language': {'name': 'go', 'version': 'go1.15.6'}}
PASSED

=============================== warnings summary ===============================
venv/lib/python3.7/site-packages/tornado/httputil.py:107
  /app/venv/lib/python3.7/site-packages/tornado/httputil.py:107: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
    class HTTPHeaders(collections.MutableMapping):

tests/agent/test_go.py: 3375 tests with warnings
  /app/venv/lib/python3.7/site-packages/tornado/httpclient.py:270: DeprecationWarning: callback arguments are deprecated, use the returned Future instead
    DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
---------- generated xml file: /app/tests/results/agent-go-junit.xml -----------
================== 4 passed, 3376 warnings, 3 rerun in 59.43s ==================
```

**Review by v1v — APPROVED** [review]
Manual testing worked like a charm, my only missing bit is to run the `make test` goal locally, but not sure if there is a missing configuration in my end. Other than that 👍

File: README.md [hunk 6]
```diff
@@ -2,28 +2,78 @@
...
+The list of HTTP methods and their purpose can be found in the [http method overview section](dyno/README.md#http-method-overview).
+arguments and expected returns, please see the in-line documentation:
+
+Endpoint|HTTP Verb|Description
+--------|---------|-----------
+/api/start|POST|Start a load-generation job
+/api/list|GET|List all configured jobs
+/api/update|POST|Updates a running job with a new configuration
+/api/stop|GET|Stops a currently-running load-generation job
+/api/scenarios|GET|Retreive a list of scenarios
```
Comment (v1v) [inline README.md:42]:
```suggestion
```

File: dyno/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+The following is a list of HTTP methods and their purpose. For specifics including possible arguments and expected returns, please see the in-line documentation:
+
+Endpoint|HTTP Verb|Description
+--------|---------|-----------
+`/api/start`|`POST`|Start a load-generation job
+`/api/list`|`GET`|List all configured jobs
+`/api/update`|`POST`|Updates a running job with a new configuration
+`/api/stop`|`GET`|Stops a currently-running load-generation job
+`/api/scenarios`|`GET`|Retreive a list of scenarios
```
Comment (v1v) [inline dyno/README.md:19]:
```suggestion
`/api/scenarios`|`GET`|Retrieve a list of scenarios
```

**Comment by beniwohli** [comment]
@cachedout ah, gotcha! Sounds like you thought of everything then :)

**Review by kuisathaverat — APPROVED** [review]
LGTM

**Review by basepi — APPROVED** [review]
Looks awesome!

**Comment by cachedout** [comment]
@v1v Do you know what's going on with this pipeline? I'm not sure what the intended behavior here but it seems like it's trying to upload an image during testing, which seems odd to me.

**Comment by v1v** [comment]
As far as I see there are two failures when running the `BATS`:

```
[2021-01-07T12:14:51.978Z] not ok 4 Test container ran OK [Procfile mode]
[2021-01-07T12:14:51.978Z] # (from function `assert_output' in file tests/test_helper/bats-assert/src/assert.bash, line 231,
[2021-01-07T12:14:51.978Z] #  in test file tests/tests.bats, line 31)
[2021-01-07T12:14:51.978Z] #   `assert_output --partial 'OK'' failed
[2021-01-07T12:14:51.978Z] # 
[2021-01-07T12:14:51.978Z] # -- output does not contain substring --
[2021-01-07T12:14:51.978Z] # substring (1 lines):
[2021-01-07T12:14:51.978Z] #   OK
[2021-01-07T12:14:51.978Z] # output (5 lines):
[2021-01-07T12:14:51.978Z] #   Starting webserver...
[2021-01-07T12:14:51.978Z] #   [2021-01-07 12:14:18 +0000] [10] [INFO] Starting gunicorn 20.0.4
[2021-01-07T12:14:51.978Z] #   [2021-01-07 12:14:18 +0000] [10] [INFO] Listening at: http://0.0.0.0:8000 (10)
[2021-01-07T12:14:51.978Z] #   [2021-01-07 12:14:18 +0000] [10] [INFO] Using worker: eventlet
[2021-01-07T12:14:51.978Z] #   [2021-01-07 12:14:18 +0000] [13] [INFO] Booting worker with pid: 13

[2021-01-07T12:16:53.793Z] not ok 8 Test container ran OK [Dyno mode]
[2021-01-07T12:16:53.793Z] # (from function `assert_output' in file tests/test_helper/bats-assert/src/assert.bash, line 231,
[2021-01-07T12:16:53.794Z] #  in test file tests/tests.bats, line 54)
[2021-01-07T12:16:53.794Z] #   `assert_output --partial 'OK'' failed
[2021-01-07T12:16:53.794Z] # 
[2021-01-07T12:16:53.794Z] # -- output does not contain substring --
[2021-01-07T12:16:53.794Z] # substring (1 lines):
[2021-01-07T12:16:53.794Z] #   OK
[2021-01-07T12:16:53.794Z] # output (5 lines):
[2021-01-07T12:16:53.794Z] #   Starting webserver...
[2021-01-07T12:16:53.794Z] #   [2021-01-07 12:16:18 +0000] [10] [INFO] Starting gunicorn 20.0.4
[2021-01-07T12:16:53.794Z] #   [2021-01-07 12:16:18 +0000] [10] [INFO] Listening at: http://0.0.0.0:8000 (10)
[2021-01-07T12:16:53.794Z] #   [2021-01-07 12:16:18 +0000] [10] [INFO] Using worker: eventlet
[2021-01-07T12:16:53.794Z] #   [2021-01-07 12:16:18 +0000] [13] [INFO] Booting worker with pid: 13
```


It's looking for the `OK` when running those docker containers, if the behaviour has changed, then the substring might need to be updated

**Comment by cachedout** [comment]
@v1v Ah, I broke a test then. I'll fix it and then ping you with the error that the pipeline gives after that point.

**Comment by cachedout** [comment]
@v1v OK, test fixed and it is now failing where it was previously. Could you take another look please and let me know if you can explain why this is trying to push to a Docker repo during the test execution?

**Pull request merged**